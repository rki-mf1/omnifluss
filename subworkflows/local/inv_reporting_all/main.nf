include { INV_REPORT_RMARKDOWN } from '../../../modules/local/inv_report_rmarkdown/main'

workflow INV_REPORTING_ALL {
    take:
    reporting_script
    fastp_jsons
    kraken_reports
    mapping_references
    markduplicate_metrics
    bedtools_genomecov
    samtools_coverage
    samtools_flagstat
    consensus_calls
    empty_kraken2_reads
    empty_spa_files

    outdir
    main:
    report                         = Channel.empty()
    versions                       = Channel.empty()
    read_statistics                = Channel.empty() //from here on, optional files that are only produced if the necessary input is available
    kraken_classification          = Channel.empty()
    mapping_statistics             = Channel.empty()
    top5_references                = Channel.empty()
    N_content_and_Ambigiuous_calls = Channel.empty()

    INV_REPORT_RMARKDOWN(
        reporting_script,
        fastp_jsons,
        kraken_reports,
        mapping_references,
        markduplicate_metrics,
        bedtools_genomecov,
        samtools_coverage,
        samtools_flagstat,
        consensus_calls,
        empty_kraken2_reads,
        empty_spa_files,
        outdir
    )
    report                        = INV_REPORT_RMARKDOWN.out.report
    versions                      = INV_REPORT_RMARKDOWN.out.versions
    //optional output
    read_statistics               = INV_REPORT_RMARKDOWN.out.read_statistics
    kraken_classification         = INV_REPORT_RMARKDOWN.out.kraken_classification
    mapping_statistics            = INV_REPORT_RMARKDOWN.out.mapping_statistics
    top5_references               = INV_REPORT_RMARKDOWN.out.top5_references
    N_content_and_Ambiguous_calls = INV_REPORT_RMARKDOWN.out.N_content_and_Ambiguous_calls

    emit:
    report                        = report
    versions                      = versions
    read_statistics               = read_statistics
    kraken_classification         = kraken_classification
    mapping_statistics            = mapping_statistics
    top5_references               = top5_references
    N_content_and_Ambiguous_calls = N_content_and_Ambiguous_calls

}