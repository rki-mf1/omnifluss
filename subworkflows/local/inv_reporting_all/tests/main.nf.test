nextflow_workflow {

    name "Test Subworkflow INV_REPORTING_ALL"
    script "../main.nf"
    workflow "INV_REPORTING_ALL"

    //  Add tags for the subworkflow
    tag "subworkflows"
    tag "reporting"
    tag "R"

    setup {
        def omnifluss_data_url = "https://github.com/rki-mf1/omnifluss_data/raw/refs/heads/INV_illumina/intermediate_data//ENA_data_wf.tar.gz"
        def omnifluss_data_command = ['bash', '-c', "curl -L --retry 5 --retry-delay 10 ${omnifluss_data_url} | tar xzf - -C ${launchDir}"]
        def omnifluss_data_proc = omnifluss_data_command.execute()
        omnifluss_data_proc.waitFor()

        if (omnifluss_data_proc.exitValue() != 0) {
            throw new RuntimeException("Error - failed to download input data: ${omnifluss_data_proc.err.text}")
        }
    }

    test("complete data") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                //check presence of sections
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")},
                //check snapshot of all outputs but the report (contains a date)
                { assert snapshot(workflow.out["N_content_and_Ambiguous_calls"], workflow.out["kraken_classification"], workflow.out["mapping_statistics"], workflow.out["read_statistics"], workflow.out["top5_references"], workflow.out["versions"]).match() }
            )
        }
    }

    //(1/8)
    test("missing fastp input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = []
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(2/8)
    test("missing kraken input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = []
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(3/8)
    test("missing kma input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = []
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(4/8)
    test("missing picard markduplicates input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = []
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(5/8)
    test("missing bedtools genomecov input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = []
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(6/8)
    test("missing samtools coverage input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = []
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(7/8)
    test("missing samtools flagstat input") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = []
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //(8/8)
    test("missing bcftools consensus data") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = []
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    test("everything missing") {
        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = []
                input[2] = []
                input[3] = []
                input[4] = []
                input[5] = []
                input[6] = []
                input[7] = []
                input[8] = []
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Read Counts</h2>") },
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Taxonomic Read Classification</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Mapping Statistics</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Coverage Distribution</h2>")},
                { assert !file(workflow.out["report"][0]).getText().contains("<h2>Consensus Sequence</h2>")}
            )
        }
    }

    //stub
    test("complete data stub") {

        tag "stub"
        options '-stub'

        when {
            workflow {
                """
                input[0] = '${projectDir}/bin/report.rmd'
                input[1] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fastp.json']).collect()
                input[2] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.report.txt']).collect()
                input[3] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.kma.spa']).collect()
                input[4] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.MarkDuplicates.metrics.txt']).collect()
                input[5] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.coverage.tsv']).collect()
                input[6] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.samtools_coverage.txt']).collect()
                input[7] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.sorted.bam.flagstat']).collect()
                input[8] = Channel.fromPath(['${launchDir}/ENA_data_wf/*.fa']).collect()
                input[9] = '.'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}