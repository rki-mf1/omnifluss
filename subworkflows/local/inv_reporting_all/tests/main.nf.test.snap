{
    "complete data": {
        "content": [
            [
                "N_content_and_Ambiguous_calls.csv:md5,7f173d58d0a2f71bde0933f29ccc8ea9"
            ],
            [
                "kraken_classification.csv:md5,e7c1324fdd1e5705b8d3f548a0cab9f3"
            ],
            [
                "mapping_statistics.csv:md5,5f50281a09ccc0373edde59061cf9ac3"
            ],
            [
                "read_statistics.csv:md5,5e9eb833f14bcd57e7404df814ac90d3"
            ],
            [
                "top5_references.csv:md5,eefb80b809c0619a29c6605ba9144661"
            ],
            [
                "versions.yml:md5,4e9198990a3bb4697a29462a7995f747"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-24T08:32:28.175734993"
    },
    "complete data stub": {
        "content": [
            {
                "0": [
                    "qc_report.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "1": [
                    "versions.yml:md5,efa6a4d53ed49017f770f004b2515348"
                ],
                "2": [
                    "read_statistics.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "kraken_classification.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "mapping_statistics.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "5": [
                    "top5_references.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "6": [
                    "N_content_and_Ambiguous_calls.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "N_content_and_Ambiguous_calls": [
                    "N_content_and_Ambiguous_calls.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "kraken_classification": [
                    "kraken_classification.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "mapping_statistics": [
                    "mapping_statistics.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "read_statistics": [
                    "read_statistics.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "report": [
                    "qc_report.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "top5_references": [
                    "top5_references.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,efa6a4d53ed49017f770f004b2515348"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-24T08:35:32.15100416"
    }
}